executor/index_lookup_hash_join.go

```diff
@@ -415,35 +409,61 @@ func (iw *indexHashJoinInnerWorker) doJoin(ctx context.Context, task *indexHashJ
 	return nil
 }
 
-func (iw *indexHashJoinInnerWorker) joinMatchedInnerRow2Chunk(ctx context.Context, innerRow chunk.Row, task *indexHashJoinTask,
-	joinResult *indexHashJoinResult, h hash.Hash64, buf []byte) (bool, *indexHashJoinResult) {
-	var err error
+func (iw *indexHashJoinInnerWorker) getMatchedOuterRows(innerRow chunk.Row, task *indexHashJoinTask, h hash.Hash64, buf []byte) (matchedRows []chunk.Row, matchedRowIdx []int, err error) {
 	h.Reset()
 	err = codec.HashChunkRow(iw.ctx.GetSessionVars().StmtCtx, h, innerRow, iw.rowTypes, iw.keyCols, buf)
 	if err != nil {
-		joinResult.err = err
-		return false, joinResult
+		return nil, nil, err
 	}
 	iw.matchedOuterPtrs = task.lookupMap.Get(h.Sum64())
 	if len(iw.matchedOuterPtrs) == 0 {
-		return true, joinResult
+		return nil, nil, nil
 	}
-	innerIter := chunk.NewIterator4Slice([]chunk.Row{innerRow})
-	var ok bool
+	matchedRows = make([]chunk.Row, 0, len(iw.matchedOuterPtrs))
+	matchedRowIdx = make([]int, 0, len(iw.matchedOuterPtrs))
 	for _, ptr := range iw.matchedOuterPtrs {
-		innerIter.Begin()
 		rowIdx := int(ptr.RowIdx)
 		outerRow := task.outerResult.GetRow(rowIdx)
-		matched, isNull, err := iw.joiner.tryToMatch(outerRow, innerIter, joinResult.chk)
+		ok, err := codec.EqualChunkRow(iw.ctx.GetSessionVars().StmtCtx, innerRow, iw.rowTypes, iw.keyCols, outerRow, iw.outerCtx.rowTypes, iw.outerCtx.keyCols)
+		if err != nil {
+			return nil, nil, err
+		}
+		if !ok {
+			continue
+		}
+		matchedRows = append(matchedRows, outerRow)
+		matchedRowIdx = append(matchedRowIdx, rowIdx)
+	}
+	return matchedRows, matchedRowIdx, nil
+}
+
+func (iw *indexHashJoinInnerWorker) joinMatchedInnerRow2Chunk(ctx context.Context, innerRow chunk.Row, task *indexHashJoinTask,
+	joinResult *indexHashJoinResult, h hash.Hash64, buf []byte) (bool, *indexHashJoinResult) {
+	matchedOuterRows, matchedOuterRowIdx, err := iw.getMatchedOuterRows(innerRow, task, h, buf)
+	if err != nil {
+		joinResult.err = err
+		return false, joinResult
+	}
+	if len(matchedOuterRows) == 0 {
+		return true, joinResult
+	}
+	var (
+		ok     bool
+		iter   = chunk.NewIterator4Slice(matchedOuterRows)
+		cursor = 0
+	)
+	for iter.Begin(); iter.Current() != iter.End(); {
+		iw.outerRowStatus, err = iw.joiner.tryToMatchOuters(iter, innerRow, joinResult.chk, iw.outerRowStatus)
 		if err != nil {
 			joinResult.err = err
 			return false, joinResult
 		}
-		if matched {
-			task.outerRowStatus[rowIdx] = outerRowMatched
-		}
-		if isNull {
-			task.outerRowStatus[rowIdx] = outerRowHasNull
+		for _, status := range iw.outerRowStatus {
+			outerRowIdx := matchedOuterRowIdx[cursor]
+			if status == outerRowMatched || task.outerRowStatus[outerRowIdx] == outerRowUnmatched {
```

Comment:
can it be replaced with the following condition?
```go
if task.outerRowStatus[outerRowIdx] == outerRowUnMatched {
    task.outerRowStatus[outerRowIdx] = status
}
```

Reply:
No
If so, when `task.outerRowStatus[outerRowIdx] == outerRowHasNull && status == outerRowMatched ` is true, it will cause error.